LightShadow: Add shadow map normal bias.

Oletus opened the pull request:
This feature offsets the position used to query shadow map occlusion along the object normal. This can be used to reduce shadow acne especially in large scenes.

Oletus committed:
Add shadow map normal offset

This feature offsets the position used to query shadow map occlusion along the object normal. This can be used to reduce shadow acne especially in large scenes.

WestLangley commented:
For reference, [here](https://digitalrune.github.io/DigitalRune-Documentation/html/3f4d959e-9c98-4a97-8d85-7a73c26145d7.htm) is a good summary of this technique -- especially as it compares to alternate approaches.

WestLangley commented on docs/api/en/renderers/WebGLRenderer.html:
```diff
@@ -249,6 +249,10 @@ <h3>[property:Integer shadowMap.type]</h3>
 
+		<h3>[property:Number shadowMap.normalOffset]</h3>
+		<p>Defines how much the position used to query shadow maps is offset along the object normal.</p>
+		<p>Increasing this value can be used to reduce shadow acne especially in large scenes. The cost is that shadows may appear distorted.</p>
```
Default is 0.

>especially in large scenes

What does "large" mean in this case?

Oletus replied:
"Large" means that there's a large shadow frustum.

Maybe something else could also be written here to explain better why not just use bias instead - bias doesn't help much in case the scene has any geometry that is at a shallow angle compared to geometry, whereas normal offset does.

WestLangley replied:
Default is *0*.

WestLangley commented:
`bias` is a property of the light in three.js. In this PR, `normalOffset` is a property of the renderer.

Would a 1x1x1 cube and a 100x100x100 cube in the same scene typically require the same value for `normalOffset`?

Oletus commented:
> Would a 1x1x1 cube and a 100x100x100 cube in the same scene typically require the same value for `normalOffset`?

@WestLangley Yes, the normal offset needed is relative to shadow map resolution and numerical precision. Increasing the size of the shadow frustum decreases shadow map precision relative to the geometry, but size of the geometry in itself doesn't have an effect on shadow map precision.

A case could be made for applying a different shadow map normal offset per light, but I wouldn't want to pay the uniform cost and the computation cost for that. Platforms like iPad/iOS can benefit greatly from this but are constrained by the number of uniform variables available.

WestLangley commented:
@Oletus 

>the normal offset needed is relative to shadow map resolution

That would naturally be a light property, then.

In the link I posted above, they claim that the bias is normalized appropriately so that the same bias automatically scales with the shadow map resolution. I assume this is the relevant [code](https://github.com/DigitalRune/DigitalRune/blob/a907452c8922a3360761e0569aa33110b0201b91/Source/DigitalRune.Graphics/Rendering/Shadows/CubeMapShadowMapRenderer.cs#L184).

That would mean one extra uniform per shadow-casting light.
```js
light.shadow.bias
light.shadow.normalOffset
```

I think I would prefer this approach, but I do not make the decision, so do what you think is best.

Thanks for the PR!  :-)

gkjohnson commented:
This is great! Thanks for making improvements to the shadows!

> A case could be made for applying a different shadow map normal offset per light, but I wouldn't want to pay the uniform cost and the computation cost for that. Platforms like iPad/iOS can benefit greatly from this but are constrained by the number of uniform variables available.

I also think it makes most sense for this field to be on the light shadow next to `.bias` because it'll be confusing if they are in two different places. If uniform count is a concern then maybe there's an argument for moving both to the renderer? If these shadow updates improve shadow quality there may not be as much of a point to hand tuning shadow values per light anyway.

I'll leave this reference from a blog post on The Witness that I came across awhile ago in case you or anyone else are interested. It describes "normal bias" as well as "slope bias" which moves the bias along the light vector as a function of normal relative to the light -- it might be able to replace our basic depth bias?:

http://the-witness.net/news/2013/09/shadow-mapping-summary-part-1/

I would also love to get rid of the fact the shadows render the opposite side for the depth map which can lead to some confusing shadow behavior and light bleeding artifacts and I think this is a good step towards that, as well!

gkjohnson commented:
Also related to #13108

Oletus commented:
Thanks for all the great feedback! I'll make the normal offset a per-light property. Performance and uniform count are minor concerns in the end, since on lower end platforms you can't run too many shadow casting lights anyway, and shadow uniforms aren't included in the shader unless lights cast shadows.

Slope scale bias could be an interesting option to explore as well, but as far as I understand it can have issues of its own in extreme cases, and could be more difficult to tune so it works well across different platforms. Normal offset is simpler and in our case it was enough to eliminate acne in a fairly challenging scene even on iPad.

Oletus commented:
Updated the PR now so that normal offset is a per light value. The vertex shader also now also computes the world normal only once and reuses it for the envmap and shadow maps.

WestLangley commented on src/renderers/shaders/ShaderChunk/envmap_vertex.glsl.js:
```diff
@@ -19,8 +19,6 @@ export default /* glsl */`
 
 		}
 
-		vec3 worldNormal = inverseTransformDirection( transformedNormal, viewMatrix );
```
>The vertex shader also now also computes the world normal only once and reuses it for the envmap and shadow maps.

For pretty-much the same reason as in #18891, I would be inclined to avoid sharing local variables between shader chunks.

This particular code block only applies to PHONG, I think.

Please reconsider creating `shadowWorldNormal` separately.... Is there a true downside to doing that?

Oletus replied:
Ah right, I thought that this code would be a part of more shaders. I agree that computing shadowWorldNormal separately makes more sense.

Oletus commented:
Put the world normal used for shadows to a separate variable again.

Ready to be merged?

Mugen87 commented:
@Oletus Can you please resolve the merge conflict in `Water.js`? Besides, it might be more safe to merge the PR with the next dev cycle `R106`.

Oletus commented:
Rebased on top of latest dev and squashed the changes.

I'd consider this a fairly low-risk change, since if normal offset is left to the default value 0 it does nothing - though of course there was a bit of reorganizing of shaders.

WestLangley commented:
>there was a bit of reorganizing of shaders

/ping @sunag

Oletus commented:
Hey, was this patch forgotten? We'd still very much like to see this merged! Ping @mrdoob :)

mrdoob commented:
Thanks!

mrdoob commented:
Now I wonder if we still need `shadow.bias`... 🤔

mrdoob commented:
@Oletus seems like `ShadowMaterial` broke:

https://raw.githack.com/mrdoob/three.js/dev/examples/physics_cannon_instancing.html
https://threejs.org/examples/physics_cannon_instancing.html

Do you mind taking a look?

Oletus commented:
Sure, I'll investigate right away.

Oletus commented:
https://github.com/mrdoob/three.js/pull/19681 fixes the issue this PR introduced in ShadowMaterial.

mrdoob commented:
@Oletus In your experience, is `shadow.bias` still relevant?

WestLangley commented:
>In your experience, is shadow.bias still relevant?

I expect it is. One is for faces largely facing the light; one for faces largely parallel to the light. See the link in https://github.com/mrdoob/three.js/pull/18915#issuecomment-600040296.

I would like it if @Oletus could explain what the units are for the bias/offset properties (as we have implemented them). Should they be normalized somehow?

See https://github.com/mrdoob/three.js/pull/18915#issuecomment-600210395.

mrdoob commented:
Seems like `normalOffset` is in world units? `bias` doesn't seem to be in world units...

WestLangley commented:
`bias` is in normalized-device-coordinate units.

It can be nice if a parameter is invariant to the shadow camera properties and/or shadow map resolution. Invariance is not required, though.

[This code](https://github.com/DigitalRune/DigitalRune/blob/a907452c8922a3360761e0569aa33110b0201b91/Source/DigitalRune.Graphics/Rendering/Shadows/CubeMapShadowMapRenderer.cs#L184) may be addressing the invariance issue, but I have not studied it.

It is nice to state what the units are in the docs.

gkjohnson commented:
> Now I wonder if we still need shadow.bias... 🤔

For reference Unity implements both "bias" and "normal bias" for shadows (reference [1](https://docs.unity3d.com/Manual/ShadowPerformance.html), [2](https://docs.unity3d.com/Manual/class-Light.html))

mrdoob commented:
Oh! `normalBias` may be a better name `normalOffset`.

@Oletus Is is okay if I rename it?

mrdoob commented:
Renamed. https://github.com/mrdoob/three.js/commit/54119884be27ab4e145318f5d858ff16aff35feb

Oletus commented:
The rename is fine by me - both are used, but maybe normal bias is easier to understand.

WestLangley commented:
@Oletus Can you please respond to https://github.com/mrdoob/three.js/pull/18915#issuecomment-645525701? Thanks.
